src/tree/updater_gpu_hist.cu

```diff
@@ -876,14 +926,22 @@ class GPUHistMaker : public TreeUpdater {
     omp_set_num_threads(nthread);
   }
 
-  bool UpdatePredictionCache(const DMatrix* data,
-                             std::vector<bst_float>* p_out_preds) override {
+  bool UpdatePredictionCache
+    (const DMatrix* data, std::vector<bst_float>* p_out_preds) override {
     return false;
   }
 
-  bool UpdatePredictionCache(
-      const DMatrix* data, HostDeviceVector<bst_float>* p_out_preds) override {
-    return false;
+  bool UpdatePredictionCache
+  (const DMatrix* data, HostDeviceVector<bst_float>* p_out_preds) override {
```

Comment:
I notice that the fast histogram updater implementation of this function has a number of checks that are lacking here:
```
// p_last_fmat_ is a valid pointer as long as UpdatePredictionCache() is called in
      // conjunction with Update().
      if (!p_last_fmat_ || !p_last_tree_ || data != p_last_fmat_) {
        return false;
      }
```
@hcho3 can you comment on the purpose of these checks and if we need them here?

It looks like they are to ensure that the cached predictions inside the current updater match the input DMatrix.

Reply:
@RAMitchell According to `tree_updater.h`, `UpdatePredictionCache` should return `false` whenever the updater lacks the capability to update the prediction cache for a given data matrix:
```
  /*!
   * \brief determines whether updater has enough knowledge about a given dataset
   *        to quickly update prediction cache its training data and performs the
   *        update if possible.
   * \param data: data matrix
   * \param out_preds: prediction cache to be updated
   * \return boolean indicating whether updater has capability to update
   *         the prediction cache. If true, the prediction cache will have been
   *         updated by the time this function returns.
   */
```
This check is essential for correct operations of the predictor. The prediction cache is a shortcut that's only valid when the updater "knows" about the dataset for which prediction is being calculated. So you cannot use the shortcut for datasets other than the training data.

So yes, you do need the check.

Reply:
Done.